File: benchmark/mandelbrot.pl
#!/usr/bin/perl -w

use strict;

my $MAXITER = 50;
my $LIMIT = 4;

sub mandelbrot($$)
{
  my $w = $_[0];
  my $h = $_[1];
  my $xmin = -1.5;
  my $ymin = -1;
  my $invN = 2/$w;

  my $checknext=1;

  for my $y (0..$h-1) {

    my $Ci = $y * $invN + $ymin;

    X:
    for my $x (0..$w-1) {

      my $Zr = 0;
      my $Zi = 0;
      my $Tr = 0;
      my $Ti = 0;

      my $Cr = $x * $invN + $xmin;

      if ($checknext) {

	  for (1..$MAXITER) {
	    $Zi = 2 * $Zr * $Zi + $Ci;
	    $Zr = $Tr - $Ti + $Cr;
	    $Ti = $Zi * $Zi;
	    $Tr = $Zr * $Zr;

	    if ($Tr + $Ti > $LIMIT) {
	      print "0";
	      next X;
	    }
	  }

	  print "1";
	  $checknext = 0;

      } else {

	for (1..$MAXITER) {
	  $Zi = 2 * $Zr * $Zi + $Ci;
	  $Zr = $Tr - $Ti + $Cr;
	  $Ti = $Zi * $Zi;
	  $Tr = $Zr * $Zr;
	}

	if ($Tr+$Ti < $LIMIT) {
	  print "1";
	} else { # $Tr+$Ti is a large number or overflow ('nan' or 'inf')
	  print "0";
	  $checknext = 1;
	}

      }

    }

    print "\n";
  } 
}

for(1..50) {
  mandelbrot(200, 200);
}

File: benchmark/mandelbrot.py
#!/usr/bin/python

import sys

MAXITER = 50
LIMIT = 4
out = sys.stdout;

def mandelbrot(w, h) :

  xmin = -1.5
  ymin = -1
  invN = 2.0 / w

  checknext = True

  for y in range(h) :
    Ci = y * invN + ymin
    
    for x in range(w) :
      Zr = 0.0
      Zi = 0.0
      Tr = 0.0
      Ti = 0.0
      Cr = x * invN + xmin  
      if (checknext) :
	for k in range(MAXITER) :
	  Zi = 2 * Zr * Zi + Ci
	  Zr = Tr - Ti + Cr
	  Ti = Zi * Zi
	  Tr = Zr * Zr
	  if (Tr + Ti) > LIMIT :
	    break
	if k == MAXITER :
	  out.write("1")
	else :
	  out.write("0")
	  checknext = False
      else :
	for k in range(MAXITER) :
	  Zi = 2 * Zr * Zi + Ci
	  Zr = Tr - Ti + Cr
	  Ti = Zi * Zi
	  Tr = Zr * Zr
	if (Tr + Ti) < LIMIT :
	  out.write("1")
	else :
	  out.write("0")
	  checknext = True
	  
    out.write("\n")

for i in range(50) :
  mandelbrot(200, 200)
